modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/api/DynamicUniversalApiHandlerRegistry.java

```diff
@@ -0,0 +1,82 @@
+package com.enonic.xp.portal.impl.api;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.stream.Collectors;
+
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+
+import com.enonic.xp.api.ApiDescriptor;
+import com.enonic.xp.app.ApplicationKey;
+import com.enonic.xp.page.DescriptorKey;
+import com.enonic.xp.portal.universalapi.UniversalApiHandler;
+import com.enonic.xp.security.PrincipalKey;
+import com.enonic.xp.security.PrincipalKeys;
+
+@Component(service = DynamicUniversalApiHandlerRegistry.class)
+public class DynamicUniversalApiHandlerRegistry
+{
+    private final ConcurrentMap<DescriptorKey, DynamicUniversalApiHandler> dynamicApiHandlers = new ConcurrentHashMap<>();
+
+    @Activate
+    public DynamicUniversalApiHandlerRegistry()
+    {
+
+    }
+
+    public DynamicUniversalApiHandler getApiHandler( final DescriptorKey descriptorKey )
+    {
+        return dynamicApiHandlers.get( descriptorKey );
+    }
+
+    public List<ApiDescriptor> getAllApiDescriptors()
+    {
+        return dynamicApiHandlers.values().stream().map( DynamicUniversalApiHandler::getApiDescriptor ).collect( Collectors.toList() );
+    }
+
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
+    public void addApiHandler( final UniversalApiHandler apiHandler, final Map<String, ?> properties )
+    {
+        final ApiDescriptor apiDescriptor = createDynamicApiDescriptor( properties );
+        this.dynamicApiHandlers.put( apiDescriptor.getKey(), new DynamicUniversalApiHandler( apiHandler, apiDescriptor ) );
+    }
+
+    public void removeApiHandler( final UniversalApiHandler apiHandler )
+    {
+        dynamicApiHandlers.values()
```

Comment:
this code has a very complicated rase and compares for equality classes without equals.